include/crocoddyl/core/integrator/rk4.hxx

```diff
@@ -0,0 +1,239 @@
+  if (enable_integration_) {
+    d->dx = (d->ki[0] + 2 * d->ki[1] + 2 * d->ki[2] + d->ki[3]) * time_step_ / 6;
+    differential_->get_state()->integrate(x, d->dx, d->xnext);
+    d->cost = (d->integral[0] + 2 * d->integral[1] + 2 * d->integral[2] + d->integral[3]) * time_step_ / 6;
+  } else {
+    d->dx.setZero();
+    d->xnext = x;
+    d->cost = d->differential[0]->cost;
+  }
+
+  // Updating the cost value
+  if (with_cost_residual_) {
+    d->r = d->differential[0]->r;
+  }
+}
+
+template <typename Scalar>
+void IntegratedActionModelRK4Tpl<Scalar>::calcDiff(const boost::shared_ptr<ActionDataAbstract>& data,
+                                                   const Eigen::Ref<const VectorXs>& x,
+                                                   const Eigen::Ref<const VectorXs>& u) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+  if (static_cast<std::size_t>(u.size()) != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "u has wrong dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  const std::size_t& nv = differential_->get_state()->get_nv();
+
+  boost::shared_ptr<Data> d = boost::static_pointer_cast<Data>(data);
+
+  differential_->calcDiff(d->differential[0], d->y[0], u);
+
+  d->dki_dy[0].topRightCorner(nv, nv).diagonal().array() = (Scalar)1;
+  d->dki_dy[0].bottomRows(nv).noalias() = d->differential[0]->Fx;
+  d->dyi_dx[0].diagonal().array() = (Scalar)1;
+  d->dki_dx[0] = d->dki_dy[0];
+  d->dki_du[0].bottomRows(nv).noalias() = d->differential[0]->Fu;
+
+  d->dli_dx[0] = d->differential[0]->Lx;
+  d->dli_du[0] = d->differential[0]->Lu;
+
+  d->ddli_ddx[0] = d->differential[0]->Lxx;
+  d->ddli_ddu[0] = d->differential[0]->Luu;
+  d->ddli_dxdu[0] = d->differential[0]->Lxu;
+
+  for (std::size_t i = 1; i < 4; ++i) {
+    differential_->calcDiff(d->differential[i], d->y[i], u);
+    d->dki_dy[i].topRightCorner(nv, nv).diagonal().array() = (Scalar)1;
+    d->dki_dy[i].bottomRows(nv).noalias() = d->differential[i]->Fx;
+
+    d->dyi_dx[i].noalias() = d->dki_dx[i - 1] * rk4_c_[i] * time_step_;
+    differential_->get_state()->JintegrateTransport(x, d->dx_rk4[i], d->dyi_dx[i], second);
+    differential_->get_state()->Jintegrate(x, d->dx_rk4[i], d->dyi_dx[i], d->dyi_dx[i], first, addto);
+    d->dki_dx[i].noalias() = d->dki_dy[i] * d->dyi_dx[i];
+
+    d->dyi_du[i].noalias() = d->dki_du[i - 1] * rk4_c_[i] * time_step_;
+    differential_->get_state()->JintegrateTransport(x, d->dx_rk4[i], d->dyi_du[i], second);
+    d->dki_du[i].noalias() = d->dki_dy[i] * d->dyi_du[i];
+    d->dki_du[i].bottomRows(nv) += d->differential[i]->Fu;
+
+    d->dli_dx[i].noalias() = d->differential[i]->Lx.transpose() * d->dyi_dx[i];
+    d->dli_du[i].noalias() = d->differential[i]->Lu.transpose() + d->differential[i]->Lx.transpose() * d->dyi_du[i];
```

Comment:
You need to split this operation to avoid memory allocation in matrix product, i.e.
```cpp
d->dli_du[i].noalias() = d->differential[i]->Lu.transpose();
d->dli_du[i].noalias().noalias() += d->differential[i]->Lx.transpose() * d->dyi_du[i];
```

Reply:
done